ADAL/src/cache/ios/ADKeychainTokenCache.m

```diff
@@ -182,8 +164,97 @@ -  (NSString*)sharedGroup
+        sWipeQuery = @{
+                       (id)kSecClass                : (id)kSecClassGenericPassword,
+                       (id)kSecAttrGeneric          : [s_wipeLibraryString dataUsingEncoding:NSUTF8StringEncoding],
+                       (id)kSecAttrAccessGroup      : _sharedGroup,
+                       (id)kSecAttrAccount          : @"TokenWipe",
+                       };
+    });
+    return [sWipeQuery mutableCopy];
+}
+
+- (BOOL)saveWipeTokenData:(ADAuthenticationError * __nullable __autoreleasing * __nullable)error
+{
+    NSMutableDictionary *query = [self wipeQuery];
+    
+    NSDictionary *wipeInfo = @{ @"bundleId" : [[NSBundle mainBundle] bundleIdentifier],
+                                @"wipeTime" : [NSDate date]
+                                };
+
+    NSData *wipeData = [NSKeyedArchiver archivedDataWithRootObject:wipeInfo];
+
+    OSStatus status = SecItemUpdate((CFDictionaryRef)query, (CFDictionaryRef)@{ (id)kSecValueData:wipeData  } );
+    if (status == errSecItemNotFound)
+    {
+        [query addEntriesFromDictionary: @{ (id)kSecAttrAccessible : (id)kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
+                                            (id)kSecValueData : wipeData
+                                            }];
+         
+        status = SecItemAdd((CFDictionaryRef)query, NULL);
+    }
+    
+    if(status != errSecSuccess)
+    {
+        NSString* details = [NSString stringWithFormat:@"Failed to log wipe data with error: %d", (int)status];
+        NSError* nserror = [NSError errorWithDomain:@"Could not log wipe data."
+                                               code:AD_ERROR_UNEXPECTED
+                                           userInfo:nil];
+        if (error)
+        {
+            *error = [ADAuthenticationError errorFromNSError:nserror
+                                                errorDetails:details
+                                               correlationId:nil];
+        }
+        return NO;
+    }
+    
+    return YES;
+}
+
+- (void)logWipeTokenData:(NSUUID *)correlationId
+{
+    static NSDictionary *sQuery;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        NSMutableDictionary *query = [self wipeQuery];
+        [query setObject:@(YES) forKey:(id)kSecReturnData];
+        sQuery = query;
+    });
+    
+    CFTypeRef data = nil;
+    OSStatus status = SecItemCopyMatching((CFDictionaryRef)sQuery, &data);
+    
+    if (status == errSecSuccess && data)
+    {
+        NSDictionary *wipeData = [NSKeyedUnarchiver unarchiveObjectWithData:(__bridge NSData * _Nonnull)(data)];
+        CFRelease(data);
+        
+        NSString *bundleId = wipeData[@"bundleId"];
+        NSDate *wipeTime = wipeData[@"wipeTime"];
+        
+        AD_LOG_INFO(correlationId, @"Last wiped by %@ at %@", bundleId, [ADHelpers stringFromDate:wipeTime]);
+        AD_LOG_INFO_PII(correlationId, @"Last wiped by %@ at %@", bundleId, [ADHelpers stringFromDate:wipeTime]);
```

Comment:
nit: double log is not necessary here

Reply:
This depends on how we expect log to be consumed. 
i.e./ we need to make decision based on AD_LOG_INFO_PII having everything including PII vs PII only.

In MSAL, we've made conscious decision on the prior. Let's discuss offline.